### ISSUE TITLE
Some Celery tasks work, others are NotRegistered

### ISSUE THREAD

**Issue by sergibondarenko** [celery/celery#3584]
## Checklist

- [x] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [ ] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce
*Django version*

`    1.9.7`

*Celery version:*

```
    celery==4.0.0
    django-celery==3.1.17

```

*Django project dir tree:*

```
    rocket
    ├── etl
    │   ├── etl
    │   │   ├── celery.py
    │   │   ├── __init__.py
    │   │   ├── settings
    │   │   │   ├── base.py
    │   │   │   ├── dev.py
    │   │   │   ├── __init__.py
    │   │   │   ├── production.py
    │   │   │   └── test.py
    │   │   ├── urls.py
    │   │   ├── wsgi.py
    │   ├── manage.py
    │   ├── pipeline
    │   │   ├── __init__.py
    │   │   ├── models.py
    │   │   ├── tasks.py
    │   │   ├── tests.py
    │   │   ├── urls.py
    │   │   ├── views.py
```

**etl/pipeline/views.py**

```
    from .tasks import *
    
    def get_response(request):
        result = add.delay(1, 2)
        result.get()
        result = sayhello.delay('tiger')
        result.get()

```
**etl/pipeline/tasks.py**

```
    from __future__ import absolute_import, unicode_literals
    from celery import shared_task
    
    @shared_task
    def add(x, y):
        return x + y
    
    @shared_task
    def mul(x, y):
        return x * y
    
    @shared_task
    def sayhello(name):
        return "Hello %s" % name
```

Also I tried this:

```
    from celery.decorators import task

    @task(name="sayhello")
    def sayhello(name):
        return "Hello {0}".format(name)
```

**etl/celery.py**

```
    from __future__ import absolute_import, unicode_literals
    import os
    from celery import Celery
    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'etl.settings.base')
    app = Celery('etl')
    app.config_from_object('django.conf:settings', namespace='CELERY')
    app.autodiscover_tasks()
    
    @app.task(bind=True)
    def debug_task(self):
        print('Request: {0!r}'.format(self.request))
```

**etl/__init__py**

```
    from __future__ import absolute_import, unicode_literals
    from .celery import app as celery_app
    __all__ = ['celery_app']
```

**etl/settings/base.py**

```
    ...
    CELERY_BROKER_URL = 'redis://localhost:6379'
    CELERY_RESULT_BACKEND = 'redis://localhost:6379'
    CELERY_ACCEPT_CONTENT = ['application/json']
    CELERY_TASK_SERIALIZER = 'json'
    CELERY_RESULT_SERIALIZER = 'json'
    CELERY_TIMEZONE = 'Europe/London'
    CELERY_IMPORTS = ('pipeline.tasks', )
```

## Expected behavior
Execute `res = sayhello.delay('trex'); res.get()` task and return string as result.

## Actual behavior
I follow the official [Celery Django](http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html) tutorial and the tasks I see in the example (`add, mul`) work for me perfectly. I get a correct response when I do `res = add.delay(1,2); res.get()`. 

But I get `*** NotRegistered: u'pipeline.tasks.sayhello'` when I try to execute another my task `res = sayhello.delay('trex')`. 

If I do `res = sayhello('trex')` then I can get the result by just typing `res`. But in this way, I execute the function in ornidarly, without using Celery.  

The **task works** if I run it **in the Django shell** `./manage shell` 

    >>> res = sayhello.delay('trex')
    >>> res.get()
    u'Hello trex'

So, the problem is that I can't execute `sayhello()` task from `pipeline/views.py`.

Why is that? How to run tasks correctly from `views.py`?

*The error full message:*

 ```
   [2016-11-11 10:56:09,870: ERROR/MainProcess] Received unregistered task of type u'pipeline.tasks.sayhello'.
    The message has been ignored and discarded.
    
    Did you remember to import the module containing this task?
    Or maybe you're using relative imports?
    
    Please see
    http://docs.celeryq.org/en/latest/internals/protocol.html
    for more information.
    
    The full contents of the message body was:
    '[["tiger"], {}, {"chord": null, "callbacks": null, "errbacks": null, "chain": null}]' (84b)
    Traceback (most recent call last):
      File "/home/trex/Development/Sirius/new/rocket/rocket-venv/local/lib/python2.7/site-packages/celery/worker/consumer/consumer.py", line 549, in on_task_received
        strategy = strategies[type_]
    KeyError: u'pipeline.tasks.sayhello'
```

[celery_proj_report.txt](https://github.com/celery/celery/files/585977/celery_proj_report.txt)

**Comment by mheppner** [comment]
I had to specify `app.autodiscover_tasks(force=True)` to get Celery to recognize my tasks. Otherwise the workers never found the tasks. This probably doesn't help as to why you can only run them in the shell though, but it's worth a shot.

**Comment by jakobholmelund** [comment]
You should't use django-celery anymore ! 

http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html

**Comment by ask** [comment]
1) When you start the worker with `-l info` you should see the names of all the tasks registered in the startup banner

Does the names in the banner differ from the error message?

The names must be the same, e.g. 'proj.pipeline.tasks.add' is different from 'pipeline.tasks.add'

2) Don't use django-celery (as mentioned by @jakobholmelund)

It doesn't seem like you're using it, but you can also remove it from your dependencies.

**Comment by GFite** [comment]
I am having an issue with some, but not all, tasks registering on autodiscover. Tasks are all annotated the same way using shared_task. All task files are in proj/app/tasks/*.py (they are all in the same folder). So, I don't believe it is a relative import or naming issue. I've stopped and cleared and purged rabbitmq and celery but I suspect there's a zombie process out there. I have some large message volumes so maybe a log or something is full.

I'm using:
django 1.10
celery 4.02
rabbitmq 3.5.7

In short, the list of registered tasks is only some of the task files in the proj/app/tasks folder, and they're all annotated the same way. Help is appreciated.

**Comment by mheppner** [comment]
@GFite I suggest posting this to the [celery-users](https://groups.google.com/forum/#!forum/celery-users) group first, instead of adding to this issue.

**Comment by Karuhanga** [comment]
This might hopefully help someone. 
I modified my code and neglected to restart the celery worker.
`Restarting the celery workers` helped.

**Comment by iddqd1** [comment]
It is happening still when settings is the module not settings.py
